lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -240,3 +295,482 @@ bool circt::firrtl::fromJSON(json::Value &value,
 
   return true;
 }
+
+/// Recursively walk a sifive.enterprise.grandcentral.AugmentedType to extract
+/// any annotations it may contain.  This is going to generate two types of
+/// annotations:
+///   1) Annotations necessary to build interfaces and store them at "~"
+///   2) Scattered annotations for how components bind to interfaces
+static bool parseAugmentedType(
+    MLIRContext *context, DictionaryAttr augmentedType, DictionaryAttr root,
+    llvm::StringMap<llvm::SmallVector<Attribute>> &newAnnotations,
+    StringRef companion, StringAttr name, StringAttr defName, Location loc,
+    Twine clazz, Twine path = {}) {
+
+  /// Optionally unpack a ReferenceTarget encoded as a DictionaryAttr.  Return
+  /// either a pair containing the Target string (up to the reference) and an
+  /// array of components or none if the input is malformed.  The input
+  /// DicionaryAttr encoding is a JSON object of a serialized ReferenceTarget
+  /// Scala class.  By example, this is converting:
+  ///   ~Foo|Foo>a.b[0]
+  /// To:
+  ///   {"~Foo|Foo>a", {".b", "[0]"}}
```

Comment:
Not here perhaps, but we may want to use @youngar 's DFS field numbering for types to attach to unify attaching to bundles/vectors in the generated attributes.

Reply:
This is an excellent idea. This would be a big improvement over the current storage approach and would make some methods like LowerType's [`filterAnnotations`](https://github.com/llvm/circt/blob/357d0c6db6c5671eefe165a53f1bd2c334199c82/lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp#L96) an `O(1)` index lookup.

This is numbering every node and not just leaves, yes?

I'll sync with @youngar on the status.